core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -284,6 +287,42 @@ public void run() {
     }
   }
 
+  /** Replaces the existing addresses, avoiding unnecessary reconnects. */
+  public void updateAddresses(EquivalentAddressGroup newAddressGroup) {
+    ManagedClientTransport savedTransport = null;
+    try {
+      synchronized (lock) {
+        EquivalentAddressGroup oldAddressGroup = addressGroup;
+        addressGroup = newAddressGroup;
+        if (state.getState() == READY || state.getState() == CONNECTING) {
+          SocketAddress address = oldAddressGroup.getAddresses().get(addressIndex);
+          int newIndex = newAddressGroup.getAddresses().indexOf(address);
+          if (newIndex != -1) {
+            addressIndex = newIndex;
+          } else {
+            // Forced to drop the connection
+            if (state.getState() == READY) {
+              savedTransport = activeTransport;
+              activeTransport = null;
+              addressIndex = 0;
```

Comment:
meh. I like it being adjacent the `activeTransport`/`pendingTransport = null`. I think the added verbosity makes it more readable.